src/util/codec/mysql/duration.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration as StdDuration;
+use time::{self, Tm};
+use std::cmp::Ordering;
+use std::fmt::{self, Display, Formatter};
+use std::io::{Read, Write};
+use std::{str, i64, u64};
+
+use util::codec::mysql::Result;
+use util::escape;
+
+const NANOS_PER_SEC: i64 = 1_000_000_000;
+const NANO_WIDTH: u32 = 9;
+const SECS_PER_HOUR: u64 = 3600;
+const SECS_PER_MINUTE: u64 = 60;
+
+/// `MAX_FSP` is the maximum digit of fractional seconds part.
+pub const MAX_FSP: usize = 6;
+/// `DEFAULT_FSP` is the default digit of fractional seconds part.
+/// `MySQL` use 0 as the default Fsp.
+pub const DEFAULT_FSP: usize = 0;
+
+/// `MAX_TIME_IN_SECS` is the maximum for mysql time type.
+const MAX_TIME_IN_SECS: u64 = 838 * SECS_PER_HOUR + 59 * SECS_PER_MINUTE + 59;
+
+fn check_fsp(fsp: usize) -> Result<()> {
+    if fsp > MAX_FSP {
+        return Err(box_err!("Invalid fsp {}", fsp));
+    }
+    Ok(())
+}
+
+/// Parse string as if it's a fraction part of a number and keep
+/// only `fsp` precision.
+fn parse_frac(s: &[u8], fsp: usize) -> Result<u32> {
+    if s.iter().any(|&c| c < b'0' || c > b'9') {
+        return Err(box_err!("{} contains invalid char", escape(s)));
+    }
+    let res = s.iter().take(fsp + 1).fold(0, |l, r| l * 10 + (r - b'0') as u32);
+    if s.len() > fsp {
+        if res % 10 >= 5 {
+            Ok(res / 10 + 1)
+        } else {
+            Ok(res / 10)
+        }
+    } else {
+        Ok(res * 10u32.pow((fsp - s.len()) as u32))
+    }
+}
+
+fn tm_to_secs(t: Tm) -> u64 {
+    t.tm_hour as u64 * SECS_PER_HOUR + t.tm_min as u64 * SECS_PER_MINUTE + t.tm_sec as u64
+}
+
+/// `Duration` is the type for `MySQL` time type.
+#[derive(Debug, Clone)]
+pub struct Duration {
+    pub dur: StdDuration,
+    past: bool,
```

Comment:
`negative` should be better.
`past` doesn't necessarily mean the duration is negative, for example, a duration from yesterday to today is a past duration, but it is still positive.